work/work.go

```diff
@@ -0,0 +1,191 @@
+package work
+
+import (
+	"context"
+	errs "errors"
+	"fmt"
+	"log"
+	"math/rand"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/heptio/workgroup"
+	"go.elastic.co/apm"
+	"go.elastic.co/apm/stacktrace"
+
+	"github.com/elastic/hey-apm/out"
+)
+
+type Worker struct {
+	// not to be modified concurrently
+	workgroup.Group
+
+	RunTimeout             time.Duration
+	AgentFlushTimeout      time.Duration
+	ServerUrl              string
+	SecretToken            string
+	TransactionLimit       int
+	TransactionFrequency   time.Duration
+	MaxSpansPerTransaction int
+	MinSpansPerTransaction int
+	ErrorLimit             int
+	ErrorFrequency         time.Duration
+	MaxFramesPerError      int
+	MinFramesPerError      int
+}
+
+type Report struct {
+	TransactionsSent uint64
+	SpansSent        uint64
+	ErrorsSent       uint64
+	Start            time.Time
+	End              time.Time
+	RequestErrors    uint64
+}
+
+func (w *Worker) Work() (Report, error) {
+
+	logger := out.NewApmLogger(log.New(os.Stderr, "", log.Ldate|log.Ltime|log.Lshortfile))
+	if w.RunTimeout < time.Second {
+		return Report{}, errs.New("run timeout too short")
+	}
+
+	tracer := newTracer(logger, w.AgentFlushTimeout, w.SecretToken, w.ServerUrl, w.MaxSpansPerTransaction)
+
+	w.Add(errors(throttle(w.ErrorFrequency), tracer, w.ErrorLimit, w.MinFramesPerError, w.MaxFramesPerError))
```

Comment:
The problem with doing these here is now `Work()` can't be reused to run any work configuration.  I was expecting we'd drive towards workload with more variety - different frequencies for transactions with many spans and few spans, etc.

Reply:
why not? you can have a struct defining `type <error|transaction>`, `limit`, `frequency`, `upperBound`, `lowerBound` and pass a list of those to the worker construct

Reply:
I didn't explain that well but also it might not matter much.  My point was that `Work` starts exactly one type of workload, defined by the members on `Worker` (`TransactionLimit`, `TransactionFrequency`, etc).  That said, the caller could still call `Worker{}.Add(customworkload)` and things would work.  I don't think that's a particularly nice API but I understand that it's practical.  I think a nicer API would have `Work` only manage a `Run` and return a `Report`, nothing more.

Reply:
> Work starts exactly one type of workload

Yes, but surely you can pass a list in a constructor, instead of just one, if you want?

>  I think a nicer API would have Work only manage a Run and return a Report, nothing more

That's fair, but what encapsulation/abstraction does the worker object provide (as `workgroup` is also exposed)?  is not better to have just 1 function then?

I am always confused by methods with pointer receivers that return something of the same type, and when computation happens implicitly (eg if `obj.A()` returns something different than `obj.B(), obj.A()`. That pattern forces me to read the entire class and all methods to understand what is going on.  That is essentially what I wanted to change.

Is [](https://github.com/jalvz/hey-apm/compare/custom-transport...jalvz:next-attempt) any better?